design-proposals/klip-7-connect-integration.md

```diff
@@ -0,0 +1,232 @@
+data.
+
+## Design
+
+### Architecture
+
+There are three ways to integrate KSQL with Connect:
+1. Run them separately and supply syntax to communicate between them
+1. Embed Connect into KSQL, managing individual connect workers as part of the KSQL ecosystem
+1. Package Connect and KSQL into a single JVM, but just spin them up at the top level
+
+There is some consensus that at scale KSQL and Connect should run as separate components. This
+means that the API that we design will not be able to leverage any "embedded-ness" of Connect,
+nullifying most of the benefits of the second approach. 
+
+There are still, however, benefits to packaging connect and KSQL in the same JVM for a seamless
+experience for new users. Namely, new users will only need to download a single package, spin up
+a single process and maintain only a single application. There are some (solvable) concerns with
+this approach, but for the reason outlined above, the syntax will be developed with the assumption
+that Connect is running as a separate component.
+
+**The recommendation is to support 1 & 3, tabling 2 unless a need for it comes in the future**
+
+### Syntax
+
+Analogous to DDL and DML, the Connect syntax will separate actions that have side effects with
+actions that simply declare metadata on top of existing data. This solves two issues:
+
+- A single connector can have many topics, which makes modeling them as a single statement difficult
+- A connector can be modeled in different ways (e.g. stream/table) and connector lifecycle should
+  be independent with how KSQL interprets it
+
+There are a few more minor points of consideration:
+- All "DML" commands that cause changes in connect will not be submitted to the command topic;
+  instead they will be issued on the REST server (this follows the pattern that all external-state
+  modifications are not distributed)
+- Some connectors will have KSQL-specific wrapper implementations that will allow simpler 
+  configuration and tighter integration. These connectors will also automatically issue the DDL 
+  statements for any topics that they create, when possible.
+
+**Sample Ingress**
+
+The first step is to create and start a Connector via the REST API (`POST /connectors`)
+```sql
+ksql> CREATE SOURCE CONNECTOR my-postgres-jdbc WITH ( \
+      connector.class="io.confluent.connect.jdbc.jdbcSourceConnector", \ 
+      tasks.max=1, \
+      connection.url=jdbc:postgresql://localhost:5432/my-db, \
+      ...);
+
+--------------
+SOURCE CREATED
+--------------
+```
+
+*Note: as Connect manages to simplify the configuration, whether through templates or any other
+means, we will immediately be able to leverage this simplification in KSQL's WITH clause.*
+
+Next, we provide tools to understand the status of the connector and corresponding topics hitting
+the `GET /connectors` and `GET /connectors/<name>` REST endpoints:
+```sql
+ksql> SHOW CONNECTORS;
```

Comment:
Would this show all connectors, or just those registered via KSQL? If the latter, how is that done?

Reply:
It should show all IMHO. We should avoid tracking connector state in KSQL.